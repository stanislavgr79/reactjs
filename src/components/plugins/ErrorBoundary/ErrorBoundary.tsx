import React from 'react';

interface Props {
  children: React.ReactNode;
}

interface ErrorState {
  error?: Error;
  errorInfo?: React.ErrorInfo;
  hasError?: boolean;
}

class ErrorBoundary extends React.Component<Props, ErrorState> {
  constructor(props: Props) {
    super(props);
    this.state = { hasError: false, error: undefined, errorInfo: undefined };
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo): void {
    this.setState({
      error: error,
      errorInfo: errorInfo,
      hasError: true,
    });
  }

  render() {
    if (this.state.hasError) {
      // eslint-disable-next-line no-console
      console.log(this.state.errorInfo);
      return (
        <div className="error-screen">
          <h2>An error has occured</h2>
          <details style={{ whiteSpace: 'pre-wrap' }}>
            {this.state.error && this.state.error.message.toString()}
            <br />
            {this.state.errorInfo && this.state.errorInfo.componentStack}
          </details>
        </div>
      );
    }
    return <React.Fragment>{this.props.children}</React.Fragment>;
  }
}

export default ErrorBoundary;
