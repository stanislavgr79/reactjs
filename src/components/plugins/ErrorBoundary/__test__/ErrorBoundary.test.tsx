import React from 'react';

import ErrorBoundary from '../ErrorBoundary';

import { render } from '@testing-library/react';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({ adapter: new Adapter() });

describe('ErrorBoundary', () => {
  test('should render children', () => {
    const tree = render(<ErrorBoundary>children</ErrorBoundary>);

    expect(tree.baseElement).toMatchSnapshot();
  });
});

describe('ErrorBoundary', () => {
  let ErrorBoundaryComponent: any;
  const testError = {
    name: 'test error',
    message: 'test error message',
  };

  beforeAll(() => {
    jest.spyOn(global.console, 'log');
    ErrorBoundaryComponent = shallow(
      <ErrorBoundary>
        <h1>wassup</h1>
      </ErrorBoundary>,
    );
    ErrorBoundaryComponent.instance().componentDidCatch(testError, 'oh nooos an error');
    ErrorBoundaryComponent.update();
    const tree = toJson(ErrorBoundaryComponent);
    expect(tree).toMatchSnapshot();
  });

  it('should update the state to indicate an error', () => {
    expect(ErrorBoundaryComponent.instance().state.hasError).toBeTruthy();
  });

  it('should not render the child component', () => {
    expect(ErrorBoundaryComponent.find('h1').exists()).toBeFalsy();
  });
});
