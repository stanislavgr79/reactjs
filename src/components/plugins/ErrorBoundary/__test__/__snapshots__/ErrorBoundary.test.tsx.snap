// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should render children 1`] = `
<body>
  <div>
    children
  </div>
</body>
`;

exports[`ErrorBoundary should render children 2`] = `
<div
  className="error-screen"
>
  <h2>
    An error has occured
  </h2>
  <details
    style={
      Object {
        "whiteSpace": "pre-wrap",
      }
    }
  >
    test error message
    <br />
  </details>
</div>
`;
